index the query terms from the percolator query

martijnvg opened the pull request:
By adding the query terms to the percolate document, many percolate requests can execute much quicker, because less percolator queries then need to be evaluated by the in-memory index.

Also added `realtime` option that contols whether a percolate request can access the percolator queries in realtime. By default the percolate requests are not executed in realtime. 

The query terms enhancement is only available in __near__ realtime mode and on indices created version `3.0.0` and later.

PR for #12664

clintongormley commented:
@martijnvg wondering if we should just execute percolation in real time always.  Percolation indices don't generally have high indexing rates, so there won't be many queries sitting unrefreshed.  This of course assumes that we can distinguish recently indexed but unrefreshed queries from those in the index?

martijnvg commented:
> This of course assumes that we can distinguish recently indexed but unrefreshed queries from those in the index?

Unfortunately we can't distinguish between refreshed and unrefreshed queries. What we can maybe do is keep track of two sets of percolate queries. One set for the refreshed ones and one set of queries that have not been refreshed yet. If a refresh happens we can move the percolator queries from the second set to the first set. At percolate time we then always need to check queries in the second set. 

Right now we can't intercept refresh calls, but we could add that (the same way index/create/delete calls get intercepted right now). Also we need to make sure percolator queries are not evaluated twice, (so not once for the first and second set of queries). This makes this approach a bit more trickier.

clintongormley commented:
@martijnvg so does setting the `realtime` parameter essentially disable this optimization?

martijnvg commented:
@clintongormley Yes, the realtime execute doesn't use the Lucene index. When execute in realtime we only iterate over all the in-memory parsed queries. When we execute **not** in realtime we basically run a query (based on percolate query, alias filter and now in this pr also the query terms query) and the percolator queries that are matched are evaluated whether they match with the document being percolated.

bleskes commented:
I’m wondering how important is real time percolation in the context of this speed up? Should we keep it simple and go with NRT all the time?


> On 18 Sep 2015, at 20:15, Martijn van Groningen <notifications@github.com> wrote:
> 
> @clintongormley Yes, the realtime execute doesn't use the Lucene index. When execute in realtime we only iterate over all the in-memory parsed queries. When we execute not in realtime we basically run a query (based on percolate query, alias filter and now in this pr also the query terms query) and the percolator queries that are matched are evaluated whether they match with the document being percolated.
> 
> —
> Reply to this email directly or view it on GitHub.
>

martijnvg commented:
@bleskes You mean remove the realtime percolation?

clintongormley commented:
Yeah, I'm leaning towards NRT as well.

kimchy commented on core/src/main/java/org/elasticsearch/action/percolate/MultiPercolateRequest.java:
```diff
@@ -18,7 +18,9 @@
  */
 package org.elasticsearch.action.percolate;
 
+import com.sun.org.apache.xpath.internal.operations.Bool;
```
?

martijnvg commented:
After thinking about it a bit more, I think removing the realtime aspect from the percolator makes sense for the following reasons:
* The realtime aspect only applies on modifications to the percolator queries and there aren't usually that many percolator queries compared to documents in a  regular index.
* The realtime aspect can still be achieved by executing a refresh after a percolate query has been added or modified. Either via the refresh api or the refresh that an index/update optionally can execute.
* Removing the realtime aspect from the percolate api allows to cleanup some code too.

martijnvg commented:
I updated this PR. The `realtime` option that has been added before in this PR has been removed and the percolator isn't realtime any more. (meaning that modifications to percolator queries aren't visible immediately to the percolator and a refresh needs to take place to make the changes visible to the percolator)

jpountz commented:
OMG

jpountz commented:
Open questions/notes:
 - if we move the percolator to NRT, then maybe we could have a "PercolatorQuery" that would use the extracted terms as an approximation and the memory index to "confirm" matches? This way we would reuse existing logic in Lucene when intersecting queries that match a documents with filters on the metadata?
 - maybe this way we could also use IndexSearcher directly and then things like running aggregations as part of the percolation process would almost come for free?
 - should we build a terms query rather than a boolean query to match queries, which will have the benefit of not failing if there are many terms? I'm wondering if we should also disable this optimization for documents that are absurdly large (probably not now but watch for issues once this optimization is released to the wild)
 - should we extract terms from the MemoryIndex instead of re-analysing fields since we need to build it anyway to confirm matches?

martijnvg commented:
@jpountz I like these ideas! I'll try them out.

martijnvg commented:
I updated this PR and made the suggested changes. It is a bit rough, but the changes look good and cleans things up in the percolator. As a nice side effect of the introduction of the `PercolatorQuery` is that the percolator becomes more unit testable!  I'll cleanup this PR and add the unit tests.

martijnvg commented:
I cleanup up this PR and added more unit tests. While cleaning up I changed another property of the percolator. The percolate api will no longer return all matches and instead returns 10 matches by default (the can be changed via the `size` parameter).

I also updated the query extract logic to only extract the longest term from a phrase query or bool query with no should clauses.

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryMetadataService {
```
Can you add javadocs to this class?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    public void extractQueryMetadata(Query query, List<Term> queryTerms) {
```
I think we could make it a better API by changing the signature to `List<Term> extractQueryTerms(Query query)` and throwing an exception if they can't be extracted. Then it would be the responsibility of the caller to add the `QUERY_METADATA_FIELD_PREFIX` prefix to field names?

martijnvg replied:
+1 makes sense

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+                if (longestTerm.bytes().length < term.bytes().length) {
+                    longestTerm = term;
+                }
+            }
```
Could you add a comment saying that the expectation is that longer terms are more likely to be less common? (I like this heuristic)

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+                        int currentSize = 0;
+                        for (Term term : bestClause) {
+                            currentSize += term.bytes().length;
+                        }
```
can you make it a utility method to make the code a bit easier to read? (compute the sum of term lengths)

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+                    noShouldClauses = false;
+                    break;
+                }
+            }
```
I'm not sure this logic is good if the boolean query has a mix of SHOULD and MUST/FILTER clauses? Shouldn't we check instead that there are some required clauses?

martijnvg replied:
The idea is that if there are SHOULD clauses then we need to extract all terms for all clauses and we can't do the optimisation of selecting the longest term per clause. So just checking for if there is a `SHOULD` clause should be sufficient?

jpountz replied:
For instance if you have `SHOULD(a) MUST(b)`, the SHOULD clause is going to be completely optional, so we would just need to extract `b`. With the current code, we would extract both `a` and `b`. So instead of checking if there are optional clauses, I think we should check that there are no required clauses?

martijnvg replied:
true, I completely misunderstood your comment.

On 8 October 2015 at 00:41, Adrien Grand <notifications@github.com> wrote:

> In
> core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java
> <https://github.com/elastic/elasticsearch/pull/13646#discussion_r41457315>
> :
>
> > +            Term longestTerm = terms[0];
> > +            for (Term term : terms) {
> > +                if (longestTerm.bytes().length < term.bytes().length) {
> > +                    longestTerm = term;
> > +                }
> > +            }
> > +            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + longestTerm.field(), longestTerm.bytes()));
> > +        } else if (query instanceof BooleanQuery) {
> > +            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
> > +            boolean noShouldClauses = true;
> > +            for (BooleanClause clause : clauses) {
> > +                if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
> > +                    noShouldClauses = false;
> > +                    break;
> > +                }
> > +            }
>
> For instance if you have SHOULD(a) MUST(b), the SHOULD clause is going to
> be completely optional, so we would just need to extract b. With the
> current code, we would extract both a and b. So instead of checking if
> there are optional clauses, I think we should check that there are no
> required clauses?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/13646/files#r41457315>.
>



-- 
Met vriendelijke groet,

Martijn van Groningen

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+        } else if (query instanceof ConstantScoreQuery) {
+            Query wrappedQuery = ((ConstantScoreQuery) query).getQuery();
+            extractQueryMetadata(wrappedQuery, queryTerms);
+        } else {
```
A new query that we might want to handle is "BoostQuery" (in the same way as ConstantScoreQuery)

jpountz replied:
And maybe also TermsQuery, which is quite common I think

martijnvg replied:
I wanted to do this, but the `TermsQuery` doesn't have methods to access the terms it encapsulates.

I'll add support for `BoostQuery`.

jpountz replied:
good call about TermsQuery, can you leave a comment with a "TODO"?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,151 @@
+        extractedTerms.add(new Term(QUERY_METADATA_FIELD_UNKNOWN));
+        Fields fields = MultiFields.getFields(indexReader);
+        for (String field : fields) {
+            // Ignore meta fields
```
I don't think it's correct? Otherwise we will fail on queries on the `_uid` or `_field_names` fields?

martijnvg replied:
true, I'll fix that

jpountz commented on core/src/main/java/org/elasticsearch/index/shard/TranslogRecoveryPerformer.java:
```diff
@@ -152,6 +155,7 @@ public void performRecoveryOperation(Engine engine, Translog.Operation operation
                                     .routing(create.routing()).parent(create.parent()).timestamp(create.timestamp()).ttl(create.ttl()),
                             create.version(), create.versionType().versionTypeForReplicationAndRecovery(), Engine.Operation.Origin.RECOVERY);
                     maybeAddMappingUpdate(engineCreate.type(), engineCreate.parsedDoc().dynamicMappingsUpdate(), engineCreate.id(), allowMappingUpdates);
+                    indexingService.preCreate(engineCreate);
```
was it a bug that we did not do it before?

martijnvg replied:
nope, because before in the `pre*` methods we would parse the query in order to validate that the query was valid. In the `post*UnderLock` methods we would parse the query again and add it to the map of parsed Lucene queries.

Now in the `pre*` methods we parse and extract the terms and adds these terms during the document. So if these methods aren't invoked now we end up not matching percolate queries at all.

jpountz replied:
ok I'll need to investigate a bit more to understand the potential consequences of this change :)

jpountz commented on core/src/main/java/org/elasticsearch/percolator/MultiDocumentPercolatorIndex.java:
```diff
@@ -71,7 +72,16 @@ public void prepare(PercolateContext context, ParsedDocument parsedDocument) {
+                    BooleanQuery.Builder bq = new BooleanQuery.Builder();
+                    bq.add(query, BooleanClause.Occur.MUST);
+                    bq.add(Queries.newNestedFilter(), BooleanClause.Occur.MUST_NOT);
+                    super.search(bq.build(), collector);
```
why didn't we need it before?

martijnvg replied:
before we did this in a different place. The `QueryCollector` has a bunch of if statements that added a query if the document being percolated has a nested field. I removed that and add an extra clause here. I think this is a better place, because it is an implementation detail of the percolator index and the fact that a document has nested mappings shouldn't leak to other places in the code.

jpountz replied:
ok, it just caught my attention because if you call directly lower-level methods this could be by-passed. I guess it's fine for now, thanks for the explanation

bleskes replied:
The call to preCreate skews indexing stats. You must call postCreate to decrement the counters increased in preCreate. Problem is that I don't think that indexing operations from the translog should be counted in the stats api. these would ideally report only use actions. Thinking about how to work around this, I wonder if we should have a callback for extra parsers - i.e., things that extract more info out of source and index it in lucene. It seems that this is what we're doing here?

martijnvg replied:
yes, that is what happens here. What if we add an extra method to `IndexingOperationListener` called `extract(ParsedDocument)` that delegates to the query metadata extraction logic?

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,173 @@
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        final Weight innerWeight = percolatorQueriesQuery.createWeight(searcher, needsScores);
+        return new Weight(this) {
```
since we don't need scores, I think you could directly create a ConstantScoreWeight

martijnvg replied:
The reason I kept scoring logic around here is that the `PercolatorQuery#percolatorQueriesQuery` ends up being a boolean query. In `PercolatorType#doPercolate(...)` the query passed down is a boolean query with 3 filter clauses and 1 must clause. The must clause is the query specified in the percolate api and allows the percolator queries to be scored based on the metadata of a percolator query.

jpountz replied:
ok

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,173 @@
+                }
+
+                final LeafReader leafReader = leafReaderContext.reader();
+                return new Scorer(this) {
```
similarly this could be a ConstantScoreScorer

jpountz replied:
(the constructor that takes a TwoPhaseIterator)

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,173 @@
+        BytesRef percolatorQueryId = new BytesRef(singleFieldsVisitor.uid().id());
+        Query percolatorQuery = percolatorQueries.get(percolatorQueryId);
+        if (percolatorQuery != null) {
+            Lucene.exists(percolatorIndexSearcher, percolatorQuery, collector);
```
I think the collector should be per-scorer, otherwise it's not thread-safe?

jpountz replied:
also should we call reset() before reusing it?

martijnvg replied:
I'm ok with creating an instance per scorer. The reason I though is was safe is that we don't execute a query in parallel of the segments and the `Lucene.exists(...)` method invokes the reset method of the collector.

jpountz replied:
I think it would be safer eg. if the query is reused

martijnvg replied:
cool, I realised this last night too (esp. if query is cached?), so the
collector has been moved and is now a normal variable in this method.

On 8 October 2015 at 00:46, Adrien Grand <notifications@github.com> wrote:

> In core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java
> <https://github.com/elastic/elasticsearch/pull/13646#discussion_r41457753>
> :
>
> > +                    @Override
> > +                    public long cost() {
> > +                        return approximation.cost();
> > +                    }
> > +                };
> > +            }
> > +        };
> > +    }
> > +
> > +    boolean matchDocId(int docId, LeafReader leafReader) throws IOException {
> > +        SingleFieldsVisitor singleFieldsVisitor = new SingleFieldsVisitor(UidFieldMapper.NAME);
> > +        leafReader.document(docId, singleFieldsVisitor);
> > +        BytesRef percolatorQueryId = new BytesRef(singleFieldsVisitor.uid().id());
> > +        Query percolatorQuery = percolatorQueries.get(percolatorQueryId);
> > +        if (percolatorQuery != null) {
> > +            Lucene.exists(percolatorIndexSearcher, percolatorQuery, collector);
>
> I think it would be safer eg. if the query is reused
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/13646/files#r41457753>.
>



-- 
Met vriendelijke groet,

Martijn van Groningen

martijnvg commented:
@jpountz Thanks for the review! I updated the PR based on your comments and answered your questions.

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,173 @@
+            public Explanation explain(LeafReaderContext leafReaderContext, int docId) throws IOException {
+                boolean match = matchDocId(docId, leafReaderContext.reader());
+                if (match) {
+                    return Explanation.match(getBoost(), "PercolatorQuery");
```
the score would not necessarily be equal to the boost?

martijnvg replied:
oops, that doesn't make sense, I fixed that now.

martijnvg commented:
@jpountz I updated the PR and fixed the other comments too.

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,192 @@
+     * since that those terms are likely to be the rarest. Boolean query's must_not clauses are always ignored ignored.
+     *
+     * If from part of the query, no query terms can be extracted then term extraction is stopped and
+     * an IllegalArgumentException is thrown.
```
maybe we need a more specific exception? I'm afraid that if eg. one of the methods that extractQueryMetadata calls throws an IAE then term extraction will silently fail

martijnvg replied:
+1

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,192 @@
+        } else if (terms2 == null) {
+            return terms1;
+        } else {
+            int terms1SumTermLength = computeSumOfTermLength(terms1);
```
maybe an average would make more sense? otherwise this would make clauses that have lots of terms too likely to be picked? eg. if the query was `foo AND (b OR a OR r OR z)` then we would index `(b,a,r,z)` while `foo` would probably be a better decision? Or maybe the clause whose shortest term is longer? (sorry I'm thinking out loud)

martijnvg replied:
> Or maybe the clause whose shortest term is longer?

I like that heuristic.

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,192 @@
+     * If from part of the query, no query terms can be extracted then term extraction is stopped and
+     * an IllegalArgumentException is thrown.
+     */
+    public List<Term> extractQueryMetadata(Query query) {
```
should this method return a set instead?

martijnvg replied:
+1

jpountz commented on core/src/test/java/org/elasticsearch/percolator/PercolatorQueryTests.java:
```diff
@@ -0,0 +1,168 @@
+
+        indexWriter.close();
+        directoryReader = DirectoryReader.open(directory);
+        IndexSearcher shardSearcher =  new IndexSearcher(directoryReader);
```
Can you do "newSearcher" instead of "new IndexSearcher" in these tests so that we get an AssertingIndexSearcher which might catch some bugs?

jpountz commented on core/src/test/java/org/elasticsearch/index/percolator/QueryMetadataServiceTests.java:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    public void testCreateQueryMetadataQuery() throws Exception {
```
Can you do the same test case on an unsupported query?

martijnvg replied:
This test tests QMS#createMetadataQuery which just iterates over all fields in the in memory index and create a big boolean should query. It doesn't have the logic for unsupported queries that is in the #extractQueryMetadata(Query) method.

jpountz commented on core/src/main/java/org/elasticsearch/percolator/TopMatchingPercolatorType.java:
```diff
@@ -0,0 +1,187 @@
+        long count = topDocs.totalHits;
+        List<BytesRef> matches = new ArrayList<>(topDocs.scoreDocs.length);
+        float[] scores = new float[topDocs.scoreDocs.length];
+        List<Map<String, HighlightField>> hls = null;
```
could we initialize it to Collections.emptyList() and then check whether the list is empty instead of non-null so that we don't have this `if` statement to pick a collector at the end of the method?

jpountz commented on core/src/main/java/org/elasticsearch/percolator/TopMatchingPercolatorType.java:
```diff
@@ -0,0 +1,187 @@
+            for (int i = 0; i < response.matches().length; i++) {
+                float score = response.scores().length == 0 ? Float.NaN : response.scores()[i];
+                Text match = new BytesText(new BytesArray(response.matches()[i]));
+                if (!response.hls().isEmpty()) {
```
I'm wondering if there is a way we could avoid all these tests on the value of hls which make the code a bit spaghetti

martijnvg replied:
Yes... let me see what can be done here, I used move the already existing code to its own class.

martijnvg replied:
Ok, a big chuck of this method can be replaced by `TopDocs.merge(...)`

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,223 @@
+
+                    @Override
+                    public int advance(int target) throws IOException {
+                        for (int docId = approximation.advance(target); docId < NO_MORE_DOCS; docId = approximation.advance(docId + 1)) {
```
could be slightly more efficient by doing:
`for (int docId = approximation.advance(target); docId < NO_MORE_DOCS; docId = approximation.nextDoc()) {`

martijnvg commented:
I was able to remove most of the custom/complex reduce logic with `TopDocs#merge`.

I'm removing the `v2.1.0` label as there are many breaking changes and the for latest commit I intentionally didn't add bwc logic. Once this PR has landed in master, I can open another PR to merge part of this change in 2.x with the bwc logic if needed.

jpountz commented on core/src/test/java/org/elasticsearch/percolator/PercolatorQueryTests.java:
```diff
@@ -0,0 +1,166 @@
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+
+public class PercolatorQueryTests extends ESTestCase {
```
maybe we could add a duel between the new fast option and the old way to make sure we don't miss some matching queries?

jpountz commented on core/src/main/java/org/elasticsearch/action/percolate/PercolateShardResponse.java:
```diff
@@ -78,39 +72,27 @@ public PercolateShardResponse(BytesRef[] matches, List<Map<String, HighlightFiel
-    }
-
-    public PercolateShardResponse(PercolateContext context, ShardId shardId) {
-        this(EMPTY_MATCHES, EMPTY_HL, 0, EMPTY_SCORES, context, shardId);
```
thanks for reducing the number of constructors

jpountz replied:
maybe we remove even more of them?

jpountz commented on core/src/main/java/org/elasticsearch/action/percolate/PercolateShardResponse.java:
```diff
@@ -78,39 +72,27 @@ public PercolateShardResponse(BytesRef[] matches, List<Map<String, HighlightFiel
     }
 
-    public List<Map<String, HighlightField>> hls() {
+    public Map<Integer, Map<String, HighlightField>> hls() {
```
can you add javadocs to say that eg. the integer key of the map represents a doc ID?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,211 @@
+/**
+ * Utility to extract query terms from queries and create queries from documents.
+ */
+public class QueryMetadataService {
```
Since we don't have state to mutate, could we make it just a helper class with a private constructor and static methods?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,211 @@
+     * an UnsupportedQueryException is thrown.
+     */
+    public Set<Term> extractQueryMetadata(Query query) {
+        // TODO: add support for the TermsQuery when it has methods to access the actual terms it encapsulates
```
let's also add a TODO for span queries?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,211 @@
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+                    if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
```
maybe we can collapse the two above if statements into: `if (clause.isRequired() == false)`?

martijnvg replied:
good catch!

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolateDocumentParser.java:
```diff
@@ -0,0 +1,234 @@
+                            doc.addDynamicMappingsUpdate(docMapper.getMapping());
+                        }
+                        if (doc.dynamicMappingsUpdate() != null) {
+                            mappingUpdatedAction.updateMappingOnMasterSynchronously(request.shardId().getIndex(), request.documentType(), doc.dynamicMappingsUpdate());
```
Do we really need to send mappings to the master node for a document that we won't index?

martijnvg replied:
The percolator had always have this "ability", percolate a skeleton document and then add the queries for that type of documents. I think we can remove this? I don't think many people are using this.

jpountz replied:
if I remember correctly, we now prevent percolator queries to be indexed if they target unmapped fields? If we actually do, then +1 to not update mappings on the master anymore.

martijnvg replied:
yes, we do that. Adding this skeleton document is just a way to add fields (automatically using the defaults) to the mapping before registering the percolator query. Instead the put mapping api should be used. I'll remove this "ability".

martijnvg replied:
There are a couple of tests that verify that this way of percolating works. The tricky (or hacky) use case is when `index.percolator.map_unmapped_fields_as_string` is enabled, in that case the check that verifies if no fields in a percolator query are unmapped is disabled. Obviously when percolating a document, the fields defined in that document don't exist too.

I think the `index.percolator.map_unmapped_fields_as_string` setting (which is kind of a hack and needs to be re-evaluated) and the lazy adding a fields via the percolate api should be addressed at the same time. But lets do that in a different issue? Otherwise this PR gets even more complex.

jpountz replied:
Works for me. Can you open an issue so that we don't forget about it?

martijnvg replied:
I've opened: #15751

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolateDocumentParser.java:
```diff
@@ -0,0 +1,234 @@
+        }
+
+        if (request.docSource() != null && request.docSource().length() != 0) {
+            doc = parseFetchedDoc(context, request.docSource(), mapperService, request.shardId().getIndex(), request.documentType());
```
should we fail if `doc` was already non-null?

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolateDocumentParser.java:
```diff
@@ -0,0 +1,234 @@
+        } finally {
+            if (parser != null) {
+                parser.close();
+            }
```
can you use try-with-resources instead?

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,223 @@
+            return super.rewrite(reader);
+        }
+
+        Query rewritten = percolatorQueriesQuery.rewrite(reader);
```
we should try to rewrite all sub queries?

jpountz replied:
woops sorry, misread

jpountz commented:
Thanks @martijnvg I think this is getting closer. However I would need help to finish the review:
 - PerFieldMappingPostingsFormat ( @rjernst maybe?) Should we have mappings for the query metadata fields or is it ok this way?
 - TranslogRecoveryPerformer ( @s1monw or @bleskes ). This code is way out of my comfort zone and I would like help confirming that changes there are ok?

martijnvg commented:
@jpountz thanks a lot for the great review. I updated the PR and addressed the comments you have made.

rjernst commented on core/src/main/java/org/elasticsearch/index/codec/PerFieldMappingPostingFormatCodec.java:
```diff
@@ -56,7 +57,11 @@ public PerFieldMappingPostingFormatCodec(Lucene50StoredFieldsFormat.Mode compres
         final MappedFieldType indexName = mapperService.indexName(field);
         if (indexName == null) {
-            logger.warn("no index mapper found for field: [{}] returning default postings format", field);
+            // the query terms are an percolator implementation detail and are added as Lucene indexed fields which are hidden
```
I'm apprehensive about this check. I think it sets a bad precedent for future changes here. Why can't the underlying fields be regular mapped fields? Do we need to better hide the percolator "type" (can we deprecate that?).

martijnvg replied:
I thought about this and my initial idea was that the percolator generates fields and the user has no control over it and therefor it should be hidden. But regardless of that this if statement is a hack... I will explore making these auto generated fields be regular mapped fields and that should allow for this if statement to be removed.

bleskes replied:
@martijnvg correct me if I'm wrong, but this will also remove the need for a custom listener method to parse when reading from the translog, right?

martijnvg replied:
no, we still need to intercept index and delete operations, right? This is just about the fields that the percolator generates are also available in the mapping. (so we don't have hidden fields) So that we don't need an if check in an unrelated place of the code base.

jpountz replied:
> I will explore making these auto generated fields be regular mapped fields and that should allow for this if statement to be removed.

Would it work to add these fields as part of the default percolator mapping?

martijnvg replied:
I don't think that would work, because the name of the query metadata field depends on name used in the query and we don't know this upfront.

bleskes replied:
> no, we still need to intercept index and delete operations, right? 

Yes, but we don't need to do it on the translog recovery anymore, right? we can let it do it's thing and then load the queries from the searcher as we do now (post refresh).

Another idea (separate PR, if you like it) - because we don't do real time percolation anymore is to cache the queries per segment and register the percolator as a warmer and only use the loading via the index reader. Then we don't need the bridge during indexing and the percolator at all. We could also consider to only load the queries into a cache upon use (since we now get them by a term look, we now what queries we need and parse them on demand only).

martijnvg replied:
@bleskes somehow in GH this comment is misplaced :) but I'm not sure if I follow... Although we don't do realtime percolation any more, we do need to add the extracted query terms as extra Lucene fields to the percolator document. How would we do that if don't intercept the document just before the actual indexing?

bleskes replied:
>  How would we do that if don't intercept the document just before the actual indexing?

I was commenting under the assumption that we solve this part via the mapping machinery.

martijnvg commented:
I merged master into this branch and fixed any issues that was caused from merging.

I gave the two outstanding issues (PerFieldMappingPostingsFormat and TranslogRecoveryPerformer) some more thought (based on @jpountz suggestion) and I think if query terms extracting happened from  a custom field mapper or object mapper (that can add new query metadata fields when the query field is parsed into a Lucene query) instead of from a custom `IndexingOperationListener` (in PercolatorQueriesRegistry) then the two outstanding issues would be fixed in a nice way:
* All percolator query meta data fields would be mapped fields (so the if statement in PerFieldMappingPostingsFormat can be removed) 
* The query metadata extracting would rely on `IndexingOperationListener ` (so the code that invoked the listener during recovery can be removed).

In order to activate this custom field impl, we can just change the default percolator mapping. Then can drop the `IndexingOperationListener` interface and also do what @bleskes suggested (load the queries per segment upon refresh and maybe only the used queries)

martijnvg commented:
Moving the query extraction to a custom field mapper is going to be hack, because the extracting requires query parsing to happen and that can't be done in a custom field mapper implementation, since the mapping code is isolated from all the non mapping service (which is good!). What can be done in a custom field mapper impl. is detect what fields query extraction is going add. (it can just traverse the contents of the `query` object field), at least this we can make sure those query meta fields are mapped.

The actual query extraction should happen somewhere else. It should be post document parse, but pre-indexing.

jpountz commented:
> What can be done in a custom field mapper impl. is detect what fields query extraction is going add. (it can just traverse the contents of the query object field), at least this we can make sure those query meta fields are mapped.

I'm wondering if we could change query extraction a bit in order to have a fixed mapping. For instance instead of extracting a `{term: { field: value}}` query into `_query_metadata_field: "value"`, we could do `_query_metadata: "field:value"`. This way we would only have two fields `_query_metadata` and `_query_metadata_unknown` that we could add to the default percolator template and then percolation would not have to add dynamic mappings for query metadata?

martijnvg commented:
@jpountz that does make things easier, let me try that out.

martijnvg commented:
Updated this PR:
* First commit changes the query terms extraction to store all extracted terms from all fields in a single query terms metadata indexed field like @jpountz suggested. (each term value uses its field name it was extracted from as a prefix). This looks solid and allowed me to drop the questionable if statement in `PerFieldMappingPostingsFormat` and allowed the query terms fields to be mapped fields in the mapping, since the fields are no longer dynamic.
* The second commit moves where the query terms extraction takes place from `PercolatorQueriesRegistry.RealTimePercolatorOperationListener#preIndex(...)` to a private method in `IndexShard` named `postPrepareIndex(...)` that gets invoked on each index operation and index recovery operation. I'm not sure if this is the best place, but it does solve the indexing stats issue that @bleskes mentioned when the query terms extraction for recovery took place in `RealTimePercolatorOperationListener`

rjernst commented:
The mapping changes LGTM. Thanks for removing the special logic!

bleskes commented:
@martijnvg and I chatted some more about this and we think we can make this work using a standard FieldMapper ala `FieldNamesFieldMapper`. Also, the hope is to only use one lucene field instead of two. Right now we have one field to hold query terms and another to indicate that the query is not parseable. However, since the query terms field's format is very strict, we feel we can use a magic value in it to indicate non-parseable queries.

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,222 @@
+ */
+public final class QueryMetadataService {
+
+    private static final byte COLUMN = 58;  // ':'
```
This worries me a bit since we allow any string as a field name. So should we use a control character instead, like the null byte? ('\0`) Then we could also index into the same field that terms could not be extracted by using an arbitrary field value that does not contain this separator.

martijnvg replied:
good point. I'll change this.

martijnvg commented on core/build.gradle:
```diff
@@ -106,6 +106,8 @@ if (isEclipse) {
     }
   }
+
+  testCompile 'org.elasticsearch:securemock:1.1'
```
@rjernst is this the right place to add this test dependency?

rjernst replied:
No, I think you probably mean to add it to the test-framework?

rjernst replied:
Unless you really mean for this to only be available for core tests.

martijnvg replied:
As it turns out I have this dependency in two places... here and in test-framework (which is what I intended):
https://github.com/elastic/elasticsearch/pull/13646/files#diff-830e6683f42cdce457750da0af4b29b6R35

I'll remove it here

martijnvg commented on core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java:
```diff
@@ -0,0 +1,231 @@
+                            @Override
+                            public float matchCost() {
+                                // matching here is expensive. what is a good value?
+                                return 1f;
```
@jpountz what would be a good value here?

jpountz replied:
Good question. This method needs to perform 1 stored field access and then pass the query through MemoryIndex. For the first part, we can approximate the cost with the number of bytes in a block (2^14). The second part is more tricky... maybe we can ignore it for now, as it depends on so many factors, maybe we can give it an arbitrary cost of eg. 1000?

So that would give:
```java
public static final float MATCH_COST =
    (1 << 14) // stored field access cost, approximated by the number of bytes in a block
    + 1000;   // cost of matching the query against the document, arbitrary as it would be really complex to estimate
```

martijnvg replied:
Cool, that makes sense to me and I've changed this in the pr.

martijnvg commented:
I've updated this PR:
* Merged in master.
* Moved the query extraction to a field mapper implementation.
* Added a bwc test that tests that percolator queries created on a 2.x cluster keep on working.

jpountz commented on core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java:
```diff
@@ -146,7 +148,12 @@ public static PercolateResponse reduce(PercolateRequest request, AtomicReference
+            try {
+                result = percolatorService.reduce(onlyCount, shardResults, request);
+            } catch (IOException e) {
+                throw new ElasticsearchException("error during reduce phase");
```
can you forward the cause of the exception?

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java:
```diff
@@ -0,0 +1,238 @@
+     * Extracts all query terms from the provided query and adds it to specified list.
+     *
+     * From boolean query with no should clauses or phrase queries only the the longest term are selected,
+     * since that those terms are likely to be the rarest. Boolean query's must_not clauses are always ignored ignored.
```
s/ignored ignored/ignored/

jpountz commented on core/src/main/java/org/elasticsearch/percolator/PercolateDocumentParser.java:
```diff
@@ -0,0 +1,228 @@
+            context.types(previousTypes);
+            if (parser != null) {
+                parser.close();
+            }
```
let's use the try-with-resources construct to close this parser?

jpountz commented on test/framework/build.gradle:
```diff
@@ -28,6 +28,7 @@ dependencies {
   compile "org.apache.httpcomponents:httpcore:${versions.httpcore}"
   compile "commons-logging:commons-logging:${versions.commonslogging}"
   compile "commons-codec:commons-codec:${versions.commonscodec}"
+  compile 'org.elasticsearch:securemock:1.2'
```
where do we use this?

martijnvg replied:
in PercolatorServiceTests and PercolateDocumentParserTests

jpountz replied:
I see, thanks!

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -129,9 +134,9 @@ public MapperService(IndexSettings indexSettings, AnalysisService analysisServic
-                    "}\n" +
+                        "\"type\" : \"percolator\"\n" +
+                    "},\n" +
+                    "\"" + QueryMetadataService.QUERY_METADATA_FIELD + "\" : { \"type\" : \"string\", \"index\" : \"not_analyzed\" }" +
```
now that there is a percolator type maybe we don't need it at all? Ie the percolator field mapper would manage its own sub fields like the geo point field mapper does for its lat/lon/geohash fields

martijnvg replied:
+1 that does make things cleaner

jpountz commented on core/src/test/java/org/elasticsearch/percolator/PercolatorQueryTests.java:
```diff
@@ -0,0 +1,246 @@
+                    queries,
+                    new MatchAllDocsQuery()
+            );
+            TopDocs topDocs2 = shardSearcher.search(builder2.build(), 10);
```
hmm what does it compare? It seems to do the same thing twice?

martijnvg replied:
it is sneaky... On builder1 `extractQueryMetadata()` gets invoked and on builder2 it doesn't.

The idea of this duel test is that the same results should be produced regardless whether query metadata is enabled.

jpountz replied:
ok I see it now!

jpountz commented:
Things look good to me in general, I mostly left minor comments. Since you refactored metadata extraction to only use a single field, could you check that we actually reject documents that have an empty field name at the root level so that it is impossible that an actual value gets extracted as if extraction was impossible?

martijnvg commented:
@jpountz Good point, documents that have no or an empty `query` field should be rejected.

martijnvg commented:
I've updated the PR:
* Renamed `QueryMetaDataService` to `ExtractQueryTermsService`
* Moved the string field holding the extracted terms under the percolator field.
* Brought back the unknown query field, because of an ambiguity issue. If a percolator query holds a query that can't be extracted we just added the nul code point (), but that collides with if we extract terms from an 'empty' field and value, which ES happily accepts (`{ "" : "" }`).

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java:
```diff
@@ -0,0 +1,150 @@
+            StringFieldMapper extractedTermsField = createStringFieldBuilder(EXTRACTED_TERMS_FIELD_NAME).build(context);
+            StringFieldMapper unknownQueryField = createStringFieldBuilder(UNKNOWN_QUERY_FIELD_NAME).build(context);
+            context.path().remove();
+            return new PercolatorFieldMapper(name(), fieldType(), fieldType(), context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo, queryShardContext, extractedTermsField, unknownQueryField);
```
the 2nd fieldType should be the default type, not the current one

jpountz commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java:
```diff
@@ -0,0 +1,150 @@
+
+        @Override
+        public Builder parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            return new Builder(parserContext.queryShardContext());
```
can you fail if node is not empty?

martijnvg replied:
It always hold one key (the field type), but I can make it fail if it contains anything other then `type`?

jpountz commented:
Only 2 minor comments left, otherwise LGTM.

jpountz commented:
As a follow-up, we should add percolator queries to our bw-testing indices (for OldIndexBackwardsCompatibilityIT). I think this is especially important now that the percolator has some bw logic.

martijnvg commented:
> As a follow-up, we should add percolator queries to our bw-testing indices

In this PR I did change the `PercolatorBackwardsCompatibilityIT` test, is that sufficient?

jpountz commented:
> In this PR I did change the PercolatorBackwardsCompatibilityIT test, is that sufficient?

What I like about OldIndexBackwardsCompatibilityIT is that it actually tests old indices instead of just simulating old indices by setting the created version manually.

jpountz commented:
LGTM

martijnvg commented:
> What I like about OldIndexBackwardsCompatibilityIT is that it actually tests old indices instead of just simulating old indices by setting the created version manually.

That is what this test did before, but now it really reads a 2.0.0 data directory and check if everything still works.

jpountz commented:
Thanks I had missed it. Nevermind then!
